### ISSUE TITLE
External monitor issue

### ISSUE THREAD

**Issue by fm4d** [ch11ng/exwm#57]
Hello, I cannot make my external monitor setup work. I have These lines in my .emacs (monitor listed as HDMI3 in xrandr):
```lisp
(require 'exwm-randr)
(setq exwm-randr-workspace-output-plist '(0 "HDMI3"))
(exwm-randr-enable)
```
I do ```xrandr --output HDMI3 --left-of LVDS1 --auto``` (laptop is listed as LVDS1 in xrandr), 
xrandr create this setup correctly (I can move my mouse on external monitor) but workspace 0 is still on laptop screen. I hope I didnt miss some instruction or so.. 

Thanks for help and thanks for awesome project.

BTW: IRC channel for exwm users would be awesome, not sure if there is enought users but still maybe worth it.

**Comment by ch11ng** [comment]
Could you provide the output of `xrandr`? And what if you evaluate `(exwm-randr--refresh)` manually?

> BTW: IRC channel for exwm users would be awesome, not sure if there is enought users but still maybe worth it.

This has been mentioned before. I think I will wait until it's released.

**Comment by fm4d** [comment]
Evaluating (exwm-randr--refresh) has no effect.
My xrandr output:
```
Screen 0: minimum 8 x 8, current 3286 x 1080, maximum 32767 x 32767
LVDS1 connected 1366x768+1920+0 (normal left inverted right x axis y axis) 277mm x 156mm
   1366x768      60.02*+
   1280x720      60.00  
   1024x768      60.00  
   1024x576      60.00  
   960x540       60.00  
   800x600       60.32    56.25  
   864x486       60.00  
   640x480       59.94  
   720x405       60.00  
   680x384       60.00  
   640x360       60.00  
DP1 disconnected (normal left inverted right x axis y axis)
DP2 disconnected (normal left inverted right x axis y axis)
DP3 disconnected (normal left inverted right x axis y axis)
HDMI1 disconnected (normal left inverted right x axis y axis)
HDMI2 disconnected (normal left inverted right x axis y axis)
HDMI3 connected 1920x1080+0+0 (normal left inverted right x axis y axis) 510mm x 287mm
   1920x1080     60.00*+
   1680x1050     59.88  
   1600x900      59.98  
   1280x1024     75.02    60.02  
   1280x800      59.91  
   1152x864      75.00  
   1280x720      59.97  
   1024x768      75.08    60.00  
   832x624       74.55  
   800x600       75.00    60.32  
   640x480       75.00    60.00  
   720x400       70.08  
VGA1 disconnected (normal left inverted right x axis y axis)
VIRTUAL1 disconnected (normal left inverted right x axis y axis)
```

**Comment by ch11ng** [comment]
Could you please replace your exwm-randr.el with [this one](https://gist.github.com/ch11ng/e4f47617aa61d5240026) and paste the log in `*Messages*`?

**Comment by fm4d** [comment]
I get this *Backtrace* message after replcing and exwm-randr--refresh:
```
Debugger entered--Lisp error: (wrong-type-argument (or eieio-object-p class-p) nil)
  signal(wrong-type-argument ((or eieio-object-p class-p) nil))
  slot-value(nil outputs)
  (message "[RANDR-DEBUG] outputs: %s" (slot-value (car (xcb:-+reply exwm--connection (xcb:-+request-unchecked exwm--connection (make-instance (quote xcb:randr:GetScreenResources) :window exwm--root)) nil)) (quote outputs)))
  (progn (message "[RANDR-DEBUG] outputs: %s" (slot-value (car (xcb:-+reply exwm--connection (xcb:-+request-unchecked exwm--connection (make-instance (quote xcb:randr:GetScreenResources) :window exwm--root)) nil)) (quote outputs))) (let ((--dolist-tail-- (slot-value (car (xcb:-+reply exwm--connection (xcb:-+request-unchecked exwm--connection ...) nil)) (quote outputs)))) (while --dolist-tail-- (let ((output (car --dolist-tail--))) (progn (let* ((v ...)) (eieio-oset v (quote name) (decode-coding-string ... ...))) (message "[RANDR-DEBUG] name: %s|%s (%s)" (slot-value (car ...) (quote name)) (slot-value (car ...) (quote crtc)) (slot-value (car ...) (quote connection))) (if (or (/= ... xcb:randr:Connection:Connected) (= 0 ...)) (plist-put output-plist (slot-value ... ...) nil) (progn (message "[RANDR-DEBUG] %dx%d%+d%+d" ... ... ... ...) (setq geometry ... output-plist ...) (if default-geometry nil ...)))) (setq --dolist-tail-- (cdr --dolist-tail--))))))
  (let (geometry output-plist default-geometry workareas viewports) (progn (message "[RANDR-DEBUG] outputs: %s" (slot-value (car (xcb:-+reply exwm--connection (xcb:-+request-unchecked exwm--connection (make-instance ... :window exwm--root)) nil)) (quote outputs))) (let ((--dolist-tail-- (slot-value (car (xcb:-+reply exwm--connection ... nil)) (quote outputs)))) (while --dolist-tail-- (let ((output (car --dolist-tail--))) (progn (let* (...) (eieio-oset v ... ...)) (message "[RANDR-DEBUG] name: %s|%s (%s)" (slot-value ... ...) (slot-value ... ...) (slot-value ... ...)) (if (or ... ...) (plist-put output-plist ... nil) (progn ... ... ...))) (setq --dolist-tail-- (cdr --dolist-tail--)))))) (message "[RANDR-DEBUG] default-geometry: %s" default-geometry) (message "[RANDR-DEBUG] output-plist: %s" output-plist) (progn (or (<= 2 (length output-plist)) (signal (quote cl-assertion-failed) (list (quote (<= 2 (length output-plist)))))) nil) (let ((--dotimes-limit-- exwm-workspace-number) (--dotimes-counter-- 0)) (while (< --dotimes-counter-- --dotimes-limit--) (let ((i --dotimes-counter--)) (let* ((output (plist-get exwm-randr-workspace-output-plist i)) (geometry (lax-plist-get output-plist output)) (frame (elt exwm-workspace--list i))) (if geometry nil (setq geometry default-geometry output nil)) (set-frame-parameter frame (quote exwm-randr-output) output) (set-frame-parameter frame (quote exwm-geometry) geometry) (progn (xcb:-+request exwm--connection (make-instance ... :window ... :value-mask ... :x ... :y ... :width ... :height ...)) (setq workareas (nconc workareas ...) viewports (nconc viewports ...))))) (setq --dotimes-counter-- (1+ --dotimes-counter--)))) (xcb:-+request exwm--connection (make-instance (quote xcb:ewmh:set-_NET_WORKAREA) :window exwm--root :data (vconcat workareas))) (xcb:-+request exwm--connection (make-instance (quote xcb:ewmh:set-_NET_DESKTOP_VIEWPORT) :window exwm--root :data (vconcat viewports))) (xcb:flush exwm--connection))
  exwm-randr--refresh()
  eval((exwm-randr--refresh) nil)
  eval-expression((exwm-randr--refresh) nil)
  call-interactively(eval-expression record nil)
  command-execute(eval-expression record)
  execute-extended-command(nil "eval-expression")
  smex-read-and-run(("eval-expression" "eshell" "load-file" "exwm-reset" "paradox-list-packages" "multi-term" "load-theme" "moe-dark" "cd" "5x5" "arp" "dbx" "dig" "erc" "ert" "eww" "ftp" "gdb" "irc" "jdb" "man" "mpc" "pdb" "pwd" "rsh" "sdb" "xdb" "calc" "diff" "dirs" "ffap" "gnus" "grep" "help" "ielm" "info" "life" "mail" "mpuz" "ping" "pong" "smex" "talk" "term" "undo" "yank" "zone" "align" "chmod" "debug" ...))
  smex()
  call-interactively(smex nil nil)
  command-execute(smex)
```

**Comment by ch11ng** [comment]
Weird. It seems RandR does not work at all. Did you get any other errors? Is EXWM working without RandR enabled? And the full content of `*Messages*`?

**Comment by fm4d** [comment]
Well, there is not really anything connected with EXWM in ```*Messages*```. EXWM is working well.. I have Emacs 24.5.1, Archlinux.

**Comment by ch11ng** [comment]
When EXWM starts up, it calls `exwm-randr--refresh`, so there should be some output unless there is some error (which should also get logged). Could you double check that? And is `exwm-randr--init` in `exwm-init-hook`?

**Comment by fm4d** [comment]
exwm-randr--refresh output in ```*Messages*```
```
[RANDR-DEBUG] outputs: (66 67 68 69 70 71 72 73 74)
[RANDR-DEBUG] name: (76 86 68 83 49)|63 (0)
[RANDR-DEBUG] 1366x768+1920+0
[RANDR-DEBUG] name: (68 80 49)|0 (1)
[RANDR-DEBUG] name: (68 80 50)|0 (1)
[RANDR-DEBUG] name: (68 80 51)|0 (1)
[RANDR-DEBUG] name: (72 68 77 73 49)|0 (1)
[RANDR-DEBUG] name: (72 68 77 73 50)|0 (1)
[RANDR-DEBUG] name: (72 68 77 73 51)|64 (0)
[RANDR-DEBUG] 1920x1080+0+0
[RANDR-DEBUG] name: (86 71 65 49)|0 (1)
[RANDR-DEBUG] name: (86 73 82 84 85 65 76 49)|0 (1)
[RANDR-DEBUG] default-geometry: [object xcb:RECTANGLE xcb:RECTANGLE t 1920 0 1366 768]
[RANDR-DEBUG] output-plist: ((76 86 68 83 49) [object xcb:RECTANGLE xcb:RECTANGLE t 1920 0 1366 768] (68 80 49) nil (68 80 50) nil (68 80 51) nil (72 68 77 73 49) nil (72 68 77 73 50) nil (72 68 77 73 51) [object xcb:RECTANGLE xcb:RECTANGLE t 0 0 1920 1080] (86 71 65 49) nil (86 73 82 84 85 65 76 49) nil)
0 (#o0, #x0, ?\C-@)
```

**Comment by ch11ng** [comment]
Hmm, could you copy Line 66 and insert it after Line 63? The output should be helpful.

**Comment by ch11ng** [comment]
All right, I have located the bug, and I will fixed shortly. Thank you for you cooperation!

**Comment by ch11ng** [comment]
I think it has been fixed in 576a676f1f0895bd473d54c2713ee9e2423023e6.

**Comment by fm4d** [comment]
Yes, fixed. Thanks :)
